pallets/parachain-staking/src/migrations.rs

```diff
@@ -0,0 +1,62 @@
+// Copyright 2019-2021 PureStake Inc.
+// This file is part of Moonbeam.
+
+// Moonbeam is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// Moonbeam is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Moonbeam.  If not, see <http://www.gnu.org/licenses/>.
+
+//! # Parachain Staking Migrations
+use crate::{
+	Config,
+	pallet::{
+		NominatorState, Nominator2, BalanceOf, NominatorState2, ExitQueue, ExitQueue2, ExitQ,
+		OrderedSet,
+	},
+};
+use sp_runtime::Perbill;
+use frame_support::weights::Weight;
+use sp_std::prelude::*;
+
+/// Storage migration for delaying nomination exits and revocations
+pub fn delay_nominator_exits_migration<T: Config>() -> (Perbill, Weight) {
+
+	// note on using pallet_migrations: migrations are not expected to be idempotent but the pallet
+	// itself will ensure that they are called only once.
```

Comment:
How is this happening in this implementation? The event and the bool storage value in the `lib.rs` are commented out on this branch.

Reply:
`pallet-migrations` has a storage map with an entry for each migration it encounters. This storage item includes enough info (currently trivial) to know (1) that it has encountered the migration before and (2) whether or not it is finished:

https://github.com/PureStake/moonbeam/blob/8ed861f7c7f48f706d5cbf6843126a1ade14fcdd/pallets/migrations/src/lib.rs#L134

It also emits events when it starts/makes progress/completes a particular migration:

https://github.com/PureStake/moonbeam/blob/8ed861f7c7f48f706d5cbf6843126a1ade14fcdd/pallets/migrations/src/lib.rs#L83

Reply:
The comment I was referring to says 

> note on using pallet_migrations: migrations are not expected to be idempotent but the pallet itself will ensure that they are called only once

Does this mean (1) `pallet-migrations` will ensure the migration is only run once or (2) the pallet with the migration must ensure it only runs once? I read it as (2) and the original question was how `parachain-staking` ensures the migration is only run once with the changes made in this PR.